src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+#include <iostream>
+#include <utility>
+#include <vector>
+
+#include "async_wrap.h"
+#include "env.h"
+#include "env-inl.h"
+#include "node_buffer.h"
+#include "node_external_reference.h"
+#include "node.h"
+#include "util.h"
+#include "zip.h"
+#include "v8.h"
+
+namespace node {
+
+namespace zip {
+
+static char const emptyZipArchive[] = {
+  0x50, 0x4B, 0x05, 0x06,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00
+};
+
+class ZipArchive final : public AsyncWrap {
+ public:
+  enum InternalFields {
+    kInternalFieldCount = AsyncWrap::kInternalFieldCount,
+  };
+
+  ZipArchive(Environment* env, v8::Local<v8::Object> obj, std::vector<char>&& buf);
+  ~ZipArchive();
+
+  void MemoryInfo(MemoryTracker* tracker) const override;
+
+  static void CheckZip(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip);
+  static void CheckZipEntry(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip, zip_uint32_t& index);
+
+  static void GetEntries(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void AddEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void DeleteEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void ReadEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void StatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void RestatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Digest(const v8::FunctionCallbackInfo<v8::Value>& args);
+
+  int GetFileType(zip_uint64_t index) const;
+
+  std::vector<char> buf_;
+  zip_error_t error_;
+  zip_source_t* source_ = nullptr;
+  zip_t* zip_ = nullptr;
+
+  SET_SELF_SIZE(ZipArchive)
+};
+
+ZipArchive::ZipArchive(Environment* env, v8::Local<v8::Object> obj, std::vector<char>&& buf)
+    : AsyncWrap(env, obj, AsyncWrap::PROVIDER_FILEHANDLE)
+    , buf_(std::move(buf)) {
+  CHECK_NE(source_ = zip_source_buffer_create(buf_.data(), buf_.size(), 0, &error_), nullptr);
+  CHECK_NE(zip_ = zip_open_from_source(source_, 0, &error_), nullptr);
+}
+
+ZipArchive::~ZipArchive() {
+  if (zip_) {
+    CHECK_EQ(zip_close(zip_), 0);
+    zip_ = nullptr;
+  }
+
+  if (source_) {
+    CHECK_EQ(zip_source_close(source_), 0);
+    zip_source_free(source_);
+  }
+}
+
+void ZipArchive::MemoryInfo(MemoryTracker* tracker) const {
+}
```

Comment:
(might want to fill this out)

Reply:
How precise does it have to be? I updated the code to track the size of the input buffer + the buffer of any file that gets added later, but it doesn't include the small-ish libzip overhead, and gets confused if the same file is modified multiple times. If the value is indicative it might be fine?

Reply:
It doesn’t have to be super precise, but it should be usable for debugging. You’ll probably want to track `buf_` here via `tracker->TrackField("buf", buf_);`. If you can’t track or estimate memory owned by libzip (including memory for added entries?) then it’s probably fine to omit it, rather than to give numbers that are potentially very inaccurate (e.g. after repeated `AddEntry()` + `DeleteEntry()` calls).